diff --git a/compiler/module_cmds.m b/compiler/module_cmds.m
index b823e9e..cb267ca 100644
--- a/compiler/module_cmds.m
+++ b/compiler/module_cmds.m
@@ -578,10 +578,6 @@ invoke_system_command_maybe_filter_output(Globals, ErrorStream, Verbosity,
         % XXX can't use Bourne shell syntax to redirect on .NET
         % XXX the output will go to the wrong place!
         CommandRedirected = Command
-    else if use_win32 then
-        % On windows we can't in general redirect standard error in the
-        % shell.
-        CommandRedirected = Command ++ " > " ++ TmpFile
     else
         CommandRedirected =
             string.append_list([Command, " > ", TmpFile, " 2>&1"])
@@ -623,20 +619,11 @@ invoke_system_command_maybe_filter_output(Globals, ErrorStream, Verbosity,
     then
         io.make_temp(ProcessedTmpFile, !IO),
 
-        % XXX we should get rid of use_win32
-        ( if use_win32 then
-            get_system_env_type(Globals, SystemEnvType),
-            ( if SystemEnvType = env_type_powershell then
-                ProcessOutputRedirected = string.append_list(
-                    ["Get-Content ", TmpFile, " | ", ProcessOutput,
-                        " > ", ProcessedTmpFile, " 2>&1"])
-            else
-                % On windows we can't in general redirect standard
-                % error in the shell.
-                ProcessOutputRedirected = string.append_list(
-                    [ProcessOutput, " < ", TmpFile, " > ",
-                        ProcessedTmpFile])
-            )
+        get_system_env_type(Globals, SystemEnvType),
+        ( if SystemEnvType = env_type_powershell then
+            ProcessOutputRedirected = string.append_list(
+                ["Get-Content ", TmpFile, " | ", ProcessOutput,
+                 " > ", ProcessedTmpFile, " 2>&1"])
         else
             ProcessOutputRedirected = string.append_list(
                 [ProcessOutput, " < ", TmpFile, " > ",
