diff --git a/library/version_array.m b/library/version_array.m
index 2b5fb78..8f97ee2 100644
--- a/library/version_array.m
+++ b/library/version_array.m
@@ -1,11 +1,10 @@
 %---------------------------------------------------------------------------%
-% vim: ts=4 sw=4 et tw=0 wm=0 ft=mercury
+% vim: ts=4 sw=4 et wm=0 ft=mercury
 %---------------------------------------------------------------------------%
 % Copyright (C) 2004-2012 The University of Melbourne.
 % Copyright (C) 2014 The Mercury Team.
 % This file may only be copied under the terms of the GNU Library General
 % Public License - see the file COPYING.LIB in the Mercury distribution.
-% vim: ft=mercury ts=4 sw=4 et wm=0 tw=0
 %---------------------------------------------------------------------------%
 %
 % File: version_array.m.
