diff --git a/compiler/compile_target_code.m b/compiler/compile_target_code.m
index c710fcf..ec87928 100644
--- a/compiler/compile_target_code.m
+++ b/compiler/compile_target_code.m
@@ -1708,15 +1708,19 @@ file_as_new_as(FileNameA, Rel, FileNameB, !IO) :-
 
 compare_file_timestamps(FileNameA, FileNameB, MaybeCompare, !IO) :-
     io.file_modification_time(FileNameA, TimeResultA, !IO),
-    io.file_modification_time(FileNameB, TimeResultB, !IO),
-    (
-        TimeResultA = ok(TimeA),
-        TimeResultB = ok(TimeB)
-    ->
-        compare(Compare, TimeA, TimeB),
-        MaybeCompare = yes(Compare)
-    ;
-        MaybeCompare = no
+    ( if FileNameA = FileNameB then
+        MaybeCompare = ( if TimeResultA = ok(_) then yes((=)) else no )
+    else
+        io.file_modification_time(FileNameB, TimeResultB, !IO),
+        ( if
+            TimeResultA = ok(TimeA),
+            TimeResultB = ok(TimeB)
+        then
+            compare(Compare, TimeA, TimeB),
+            MaybeCompare = yes(Compare)
+        else
+            MaybeCompare = no
+        )
     ).
 
 %-----------------------------------------------------------------------------%
